File: modules/bounty.py
import universe
from go_to_adjacent_systems import go_to_adjacent_systems
from go_somewhere_significant import go_somewhere_significant
import vsrandom
import launch
import faction_ships
import Director
import Briefing
import unit
import VS
import quest
class bounty (Director.Mission):
	def SetVar (self,val):
	  if (self.var_to_set!=''):
	    quest.removeQuest (self.you.isPlayerStarship(),self.var_to_set,val)
	def __init__ (self,minnumsystemsaway, maxnumsystemsaway, creds, run_away, shipdifficulty, tempfaction,jumps=(),var_to_set=''):
	  Director.Mission.__init__ (self)
	  run_away=0
	  self.newship=""
	  self.mplay="all"
	  self.var_to_set = var_to_set
	  self.istarget=0
	  self.obj=0
	  self.curiter=0
	  self.arrived=0
	  self.faction = tempfaction	  
	  self.difficulty = shipdifficulty
	  self.runaway=run_away
	  self.cred=creds
	  mysys=VS.getSystemFile()
	  sysfile = VS.getSystemFile()
	  self.you=VS.getPlayer()
	  self.enemy=VS.Unit()
	  self.adjsys=go_to_adjacent_systems (self.you,vsrandom.randrange(minnumsystemsaway,maxnumsystemsaway+1),jumps)
	  self.mplay=universe.getMessagePlayer(self.you)
	  if (self.you):
	    VS.IOmessage (0,"bounty mission",self.mplay,"Good Day, %s. Your mission is as follows:" % (self.you.getName()))
	    self.adjsys.Print("You should start in the system named %s","Then jump to %s","Finally, jump to %s, your final destination","bounty mission",1)
	    VS.IOmessage (1,"bounty mission",self.mplay,"Once there, you must destroy a %s unit." % (self.faction))
	    VS.IOmessage (2,"bounty mission",self.mplay,"You will then %.2f credits as your reward (if you survive)." % (self.cred))
	    VS.IOmessage (3,"bounty mission",self.mplay,"#00ff00Good luck!")
	  else:
	    print "aboritng bounty constructor..."
	    VS.terminateMission (0)
	
	def Win (self,un,terminate):
	  self.SetVar(1)
	  VS.IOmessage (0,"bounty mission",self.mplay,"#00ff00Excellent work pilot.")
	  VS.IOmessage (0,"bounty mission",self.mplay,"#00ff00You have been rewarded for your effort as agreed.")
	  VS.IOmessage (0,"bounty mission",self.mplay,"#00ff00Your contribution to the war effort will be remembered.")
	  un.addCredits(self.cred)
	  if (terminate):
	    print "you win bounty mission!"
	    VS.terminateMission(1)
	  
	def Lose (self,terminate):
	  VS.IOmessage(0,"bounty mission",self.mplay,"#ff0000You have failed this mission and will not be rewarded.")
	  self.SetVar(-1)
	  if (terminate):
	    print "lose bounty mission"
	    VS.terminateMission(0)
	  
	def Execute (self):
	  isSig=0
	  if (self.you.isNull()):
	    self.Lose (1)
	    return
	  if (self.arrived==2):
	    if (not self.runaway):
	      if (not self.istarget):
		if (self.enemy):
		  curun=VS.getUnit(self.curiter)
		  self.curiter+=1
		  if (curun==self.enemy):
		    self.enemy.SetTarget(self.you)
		  elif (curun.isNull()):
		    self.curiter=0
	    if (self.enemy.isNull()):
	      self.Win(self.you,1)
	      return
	  elif (self.arrived==1):
	    significant=self.adjsys.SignificantUnit()
	    if (significant.isNull ()):
	      print "sig null"
	      VS.terminateMission(0)
	      return
	    else:
	      if (self.you.getSignificantDistance(significant)<10000.0):
		if (self.newship==""):
		  self.newship=faction_ships.getRandomFighter(self.faction)
		self.enemy=launch.launch_wave_around_unit("Shadow",self.faction,self.newship,"default",1+self.difficulty,3000.0,4000.0,significant)
		self.obj=VS.addObjective("Destroy the Shadow %s ship." % (self.enemy.getName ()))
		if (self.enemy):
		  if (self.runaway):
		    self.enemy.SetTarget(significant) #CHANGE TO SetTarget ==>NOT setTarget<==
		    self.enemy.ActivateJumpDrive(0)
		  self.arrived=2
		else:
		  print "enemy null"
		  VS.terminateMission(0)
		  return
	  else:
	    if (self.adjsys.Execute()):
	      self.arrived=1
	      self.newship=faction_ships.getRandomFighter(self.faction)
	      self.adjsys=go_somewhere_significant(self.you,0,500)
	      localdestination=self.adjsys.SignificantUnit().getName()
	      VS.IOmessage (3,"bounty mission",self.mplay,"You must destroy the %s unit in this system." % (self.newship))
	      if (self.runaway):	#ADD OTHER JUMPING IF STATEMENT CODE HERE
	        VS.IOmessage (4,"bounty mission",self.mplay,"He is running towards the jump point.  Catch him!")
	        VS.IOmessage (5,"bounty mission",self.mplay,"he is going to %s" % (localdestination))
	      else:
	        VS.IOmessage (4,"bounty mission",self.mplay,"Scanners are picking up a metallic object!")
	        VS.IOmessage (5,"bounty mission",self.mplay,"Coordinates appear near %s" % (localdestination))
	
	def initbriefing(self):
		print "ending briefing"                
	
	def loopbriefing(self):
		print "loop briefing"
		Briefing.terminate();
	
	def endbriefing(self):
		print "ending briefing"        	  
	

def initrandom (minns, maxns, credsmin, credsmax, run_away, minshipdifficulty, maxshipdifficulty,jumps=(),var_to_set=''):
  you=VS.getPlayer()
  tempfaction='aera'
  if (you):
    name = you.getFactionName ()
    factionname=vsrandom.randrange(0,faction_ships.getMaxFactions())
    tempfaction=faction_ships.intToFaction(factionname)
    i=0
    while ((name==tempfaction or name=="unknown") and i<10):
      factionname=vsrandom.randrange(0,faction_ships.getMaxFactions())
      tempfaction=faction_ships.intToFaction(factionname)
      i+=1
    sd = vsrandom.random()*(maxshipdifficulty-minshipdifficulty)+minshipdifficulty
    return bounty (minns,maxns,(1.0+(sd*0.5))*(vsrandom.random ()*(credsmax-credsmin)+credsmin),run_away,sd,tempfaction,jumps,var_to_set)
  else:
    print "aborting bounty initrandom"
    VS.terminateMission(0)
  

File: modules/bounty.c
module bounty {
	object faction;
	object destination;
	object enemycontainer;
	object youcontainer;
	object sigcont;
	object newship;
	int arrived;
	int difficulty;
	int curiter;
	int difficulty;
	float cred;
	//	import ai_stationary;
	import universe;
	import random;
	import launch;
	import faction_ships;
	bool istarget;
	bool runaway;
	object systemlist;
	void destroy () {
	  _olist.delete (systemlist);
	  _string.delete (faction);
	  _string.delete (destination);
	  if (!_std.isNull(sigcont)) {
	    _unit.deleteContainer (sigcont);
	  }
	  if (!_std.isNull(youcontainer)) {
	     _unit.deleteContainer (youcontainer);
	  }
	  if (!_std.isNull(enemycontainer)) {
	    _unit.deleteContainer (enemycontainer);
	  }
	};
	void initrandom (int minns, int maxns, float credsmin, float credsmax, bool run_away, int minshipdifficulty, int maxshipdifficulty) {
	  faction_ships.init();	    
	  object you=_unit.getPlayer();
	  object tempfaction;
	  if (!_std.isNull(you)) {
	    object name = _unit.getFaction (you);
	    int factionname=random.randomint(0,faction_ships.getMaxFactions()-1);
	    tempfaction=faction_ships.intToFaction(factionname);
	    while (_string.equal(name,tempfaction)) {
	      int factionname=random.randomint(0,faction_ships.getMaxFactions()-1);
	      tempfaction=faction_ships.intToFaction(factionname);
	    }
	    int sd = random.randomint (minshipdifficulty,maxshipdifficulty);
	    initstage2 (minns,maxns,(1.0+(_std.Float(sd)*0.5))*random.random (credsmin, credsmax),run_away,sd,tempfaction);
	    _string.delete (name);
	  }else {
	    _std.terminateMission(false);
	  }
	};
	void init (int minnumsystemsaway, int maxnumsystemsaway, float creds, bool run_away, int shipdifficulty, object tempfaction) {
	  faction_ships.init();	    
	  initstage2 (minnumsystemsaway,maxnumsystemsaway,creds,run_away,shipdifficulty,tempfaction);
	};
	void initstage2 (int minnumsystemsaway, int maxnumsystemsaway, float creds, bool run_away, int shipdifficulty, object tempfaction) {

	  systemlist = _olist.new();
	  faction = tempfaction;	  
	  _std.setNull(newship);
	  _std.setNull(sigcont);
	  _std.setNull(enemycontainer);
	  _std.setNull(youcontainer);
	  difficulty = shipdifficulty;
	  runaway=run_away;

	  arrived=0;
	  curiter=0;
	  istarget=false;
	  cred=creds;
	  object mysys=_std.getSystemFile();
	  object sysfile = _std.getSystemFile();
	  object you=_unit.getPlayer();
	  youcontainer = _unit.getContainer (you);
	  if (!_std.isNull(you)) {

	    object name=_unit.getName(you);
	   
	    object str = _string.new();
	    _io.sprintf(str,"Good Day, %s. Your mission is as follows:",name);
	    _string.delete (name);
	    _io.message (0,"game","all",str);
	    _io.message (0,"game","all","In order to get to your destination, you must:");
	    destination=universe.getAdjacentSystem (sysfile,random.randomint (minnumsystemsaway,maxnumsystemsaway),systemlist);
	    _io.sprintf(str,"Once there, you must destroy a %s unit.",faction);
	    _io.message (2,"game","all",str);
	    _io.message (3,"game","all","You will then recieve credits as your reward");
	    _io.message (4,"game","all","(if you survive).  Good luck!");
	    _string.delete(str);
	  } else {
	    _std.terminateMission (false);
	  }
	};

	void Win (object un,bool terminate) {
	  _io.message (0,"game","all","Excellent work pilot.");
	  _io.message (0,"game","all","You have been rewarded for your effort as agreed.");
	  _io.message (0,"game","all","Your contribution to the war effort will be remembered.");
	  _unit.addCredits(un,cred);
	  if (terminate) {
	    destroy();
	    _std.terminateMission(true);
	  }
	};

	void Lose (bool terminate) {
	  _io.message(0,"game","all","You have failed this mission and will not be rewarded.");
	  if (terminate) {
	    destroy();
	    _std.terminateMission(false);
	  }
	};

	void loop () {
	  bool isSig;
	  object enemy;
	  object you=_unit.getUnitFromContainer(youcontainer);
	  if (_std.isNull(you)) {
	    Lose (true);
	    return;
	  }
	  if (arrived==3) {
	    enemy=_unit.getUnitFromContainer(enemycontainer);
	    if (!istarget) {
	      object curun=_unit.getUnit(curiter);
	      if (!_std.isNull (enemy)) {
		if (_std.equal(curun,enemy)) {
		  _unit.setTarget(enemy,you);
		}
	      }
	      curiter=curiter+1;
	    }
	    if (_std.isNull(you)) {
	      Lose(true);
	      return;
	    }
	    if (_std.isNull(enemy)) {
	      Win(you,true);
	      return;
	    }
	  } else if (arrived==2) {
	    object sysfil = _std.getSystemFile();
	    if (!_string.equal (sysfil,destination)) {
	      arrived=3;
	      enemy=_unit.getUnitFromContainer(enemycontainer);
	      if (_std.isNull(you)) {
		Lose(true);
		return;
	      }
	      if (_std.isNull(enemy)) {
		Win(you,true);
		return;
	      }
	    } else {
	      _std.ResetTimeCompression();
	    }
	  } else if (arrived==1) {
	    object significant=_unit.getUnitFromContainer(sigcont);
	    if (_std.isNull (significant)) {
	      _std.terminateMission(false);
	    }else {
	      if (unit.getSignificantDistance(you,significant)<10000.0) {
		if (_std.isNull(newship)) {
		  newship=faction_ships.getRandomFighter(faction);
		}
		enemy=launch.launch_wave_around_unit("Base",faction,newship,"default",1+difficulty,3000.0,4000.0,significant);
		enemycontainer=_unit.getContainer(enemy);
		if (!_std.isNull(enemy)) {
		  if (runaway) {
		    _unit.setTarget(enemy,significant);
		    _unit.Jump(enemy);
		    arrived=2;
		  } else {
		    arrived=3;
		  }
		}
	      }
	    }
	  } else {
	    object sysfil = _std.getSystemFile();
	    if (_string.equal (sysfil,destination)) {
	      bool isSig=false;
	      arrived=1;
	      object newship=faction_ships.getRandomFighter(faction);
	      int randint=random.randomint(0,50);
	      object significant = unit.getJumpPoint (randint);

	      if (_std.isNull (significant)) {
		significant =_unit.getPlayer();
	      } else {
		sigcont=_unit.getContainer(significant);
		isSig=true;
	      }
	      if (_std.isNull(significant)) {
		a_std.terminateMission (false);
	      } else {
		object localdestination=_unit.getName(significant);
		if (isSig) {	//ADD OTHER JUMPING IF STATEMENT CODE HERE
		} else {
		  enemy=launch.launch_wave_around_unit("Base",faction,newship,"default",1+difficulty,500.0,1000.0,significant);
		  enemycontainer=_unit.getContainer(enemy);
		  _unit.setTarget(you,enemy);
		  arrived=2;
		}
		object str = _string.new();
		_io.sprintf(str,"You must destroy the %s unit in this system.",newship);
		_io.message (0,"game","all",str);
		if (isSig) {	//ADD OTHER JUMPING IF STATEMENT CODE HERE ALS0
		  if (runaway) {
		    _io.message (3,"game","all","oh no... He is running towards the jump point.  Catch him quick!");
		    _io.sprintf(str,"he is going to %s",localdestination);
		  }else {
		    _io.message (3,"game","all","Scanners are picking up a metallic object!");
		    _io.sprintf(str,"Coordinates appear near %s",localdestination);
		  }
		  _io.message (4,"game","all",str);
		}
		_string.delete(str);
		
	      }
	    }
	    _string.delete (sysfil);
	  }
	};
	void initbriefing() {
	  
	};
	void loopbriefing() {
	  
	};
	void endbriefing() {
	  
	};
}
